src/aws-cpp-sdk-core/include/aws/core/client/AWSClientEventStreamingAsyncTask.h

```diff
@@ -0,0 +1,137 @@
+                                  const HandlerT& handler, const std::shared_ptr<const Aws::Client::AsyncCallerContext>& handlerContext,
+                                  const std::shared_ptr<Utils::Event::EventEncoderStream>& stream, const Http::HttpMethod method,
+                                  char const* const signerName)
+      : m_clientThis(pClientThis),
+        m_endpoint(std::move(endpoint)),
+        m_pRequest(pRequest),
+        m_handler(handler),
+        m_handlerContext(handlerContext),
+        m_stream(stream),
+        m_method(method),
+        m_signerName(signerName),
+        m_sem(Aws::MakeShared<Aws::Utils::Threading::Semaphore>("BidirectionalEventStreamingTask", 0, 1)) {
+    assert(m_clientThis);
+    assert(m_pRequest);
+    assert(m_signerName);
+
+    // Reset EventStream handler as it still has a raw dangling pointer of the original user request
+    m_pRequest->SetEventStreamHandler(m_pRequest->GetEventStreamHandler());
+
+    // Perform synchronization logic extracted from generated code
+    // Setup a signal to notify the submitter thread that the task has actually started
+    std::shared_ptr<Aws::Utils::Threading::Semaphore> sem = m_sem;
+    m_pRequest->SetRequestSignedHandler([stream, sem](const Aws::Http::HttpRequest& httpRequest) {
+      stream->SetSignatureSeed(Aws::Client::GetAuthorizationHeader(httpRequest));
+      sem->ReleaseAll();
+    });
+
+    std::weak_ptr<RequestT> wRequest = m_pRequest;
+    // Setup InitialResponse handler to use the new actual request object
+    if (!m_pRequest->GetHeadersReceivedEventHandler()) {
+      m_pRequest->SetHeadersReceivedEventHandler([wRequest](const Http::HttpRequest*, Http::HttpResponse* response) {
+        auto request = wRequest.lock();
+        if (!request) {
+          AWS_LOGSTREAM_FATAL(ClientT::GetAllocationTag(), "Unexpected nullptr bi-directional streaming request on initial response!");
+          assert(false);
+          return;
+        }
+        if (!response) {
+          AWS_LOGSTREAM_FATAL(ClientT::GetAllocationTag(), "Unexpected response nullptr!");
+          assert(false);
+          return;
+        }
+        auto& initialResponseHandler = request->GetEventStreamHandler().GetInitialResponseCallbackEx();
+        if (initialResponseHandler) {
+          initialResponseHandler({response->GetHeaders()}, Utils::Event::InitialResponseType::ON_RESPONSE);
+        }
+      });
+    }
+
+    // Setup ResponseStreamFactory to provide EventStream decoder based on the new actual request object, not the original one.
+    m_pRequest->SetResponseStreamFactory([wRequest]() -> Aws::IOStream* {
+      auto request = wRequest.lock();
+      if (!request) {
+        AWS_LOGSTREAM_FATAL(ClientT::GetAllocationTag(),
+                            "Unexpected nullptr bi-directional streaming request on response streaming factory call!");
+        assert(false);
+        return nullptr;
```

Comment:
instead of nullptr should we return a empty string stream? even though we have entered a invalid state, a empty stream might help someone more gracefully exit than a null ptr

Reply:
ResponseStream is already handling nullptr condition in the c-tor.
https://github.com/aws/aws-sdk-cpp/blob/9c322a9857208116ee6e6f41b5483073c3fa4c73/src/aws-cpp-sdk-core/source/utils/stream/ResponseStream.cpp#L26-L30